microannotate/generator.py

```diff
@@ -198,101 +113,253 @@ def get_revs(hg, rev_start=0, rev_end="tip"):
+        self.tokenize_enabled = tokenize
+        self.remove_comments_enabled = remove_comments
+
+    async def remove_comments(self, path, content):
+        try:
+            async with self.session.post(
+                f"http://localhost:{self.code_analysis_port}/comment?file_name={path}",
+                data=content,
+            ) as r:
+                # The server returns 200 when successful, 204 when no comments have been removed and
+                # 404 when an extension is not supported.
+
+                text = await r.text()
+
+                if r.status == 200:
+                    content = text.encode("utf-8")
+                elif r.status not in [204, 404]:
+                    logger.error(
+                        f"Error {r.status} from the code analysis server, for {path}: {text}"
+                    )
+        except aiohttp.ClientConnectionError as e:
+            logger.error(f"Error connecting to code analysis server, for {path}: {e}")
+
+        return content
+
+    async def write_file(self, commit, path):
+        loop = asyncio.get_event_loop()
+
+        try:
+            content = await loop.run_in_executor(None, _hg_cat, path, commit.node)
+        except hglib.error.CommandError as e:
+            if b"no such file in rev" in e.err:
+                # The file was removed.
+                os.remove(os.path.join(self.repo.workdir, path.decode("ascii")))
+                self.repo.index.remove(path)
+                return
+            else:
+                raise
+
+        path = path.decode("ascii")
+
+        os.makedirs(
+            os.path.dirname(os.path.join(self.repo.workdir, path)), exist_ok=True
+        )
+
+        if self.remove_comments_enabled:
+            content = await self.remove_comments(path, content)
+
+        async with aiofiles.open(os.path.join(self.repo.workdir, path), "wb") as f:
+            if self.tokenize_enabled:
+                await f.writelines(
+                    word.group(0) + b"\n" for word in SPLIT_WORD_REGEX.finditer(content)
                 )
-            except FileNotFoundError:
-                raise Exception("rust-code-analysis is required for comment removal")
+            else:
+                await f.write(content)
 
-            for _ in range(7):
-                try:
-                    r = requests.get(f"http://localhost:{code_analysis_port}/ping")
-                    r.raise_for_status()
-                    ready = True
-                    break
-                except Exception:
-                    if proc.poll() is not None:
+        self.repo.index.add(path)
+
+    async def convert(self, commit):
+        set_modified_files(commit)
+
+        copy_target_paths = set(commit.file_copies.values())
+
+        before_after_paths = []
+        for path in commit.files:
+            if path in commit.file_copies:
+                before_after_paths.append((path, commit.file_copies[path]))
+            elif path in copy_target_paths:
```

Comment:
nit: "not in" to avoid "pass" and an other "else"

Reply:
Fixed.